types/prop-types/prop-types-tests.ts

```diff
@@ -159,3 +159,49 @@ type UnmatchedPropKeys2 = Pick<ExtractedProps, Extract<{
 }[keyof ExtractedProps], keyof ExtractedProps>>;
 
 PropTypes.checkPropTypes({ xs: PropTypes.array }, { xs: [] }, 'location', 'componentName');
+
+// This would be the type that JSX sees
+type Defaultize<T, D> =
+    & Pick<T, Exclude<keyof T, keyof D>>
+    & Partial<Pick<T, Extract<keyof T, keyof D>>>
+    & Partial<Pick<D, Exclude<keyof D, keyof T>>>;
+
+// This would be the type inside the component
+type Undefaultize<T, D> =
+    & Pick<T, Exclude<keyof T, keyof D>>
+    & { [K in Extract<keyof T, keyof D>]-?: NonNullable<T[K]>; }
```

Comment:
I'm waffling on this one. `PropTypes.string` (for example) allows for a nullable string, but if the user passes in `undefined` explicitly for a prop with that prop type (or leaves it out), the default props kick in, but if they pass in `null` explicitly, the default props do not kick in. So really this should be `Exclude<T[K], undefined>` instead of also excluding null, but I'm also fairly confident that people don't actually do this IRL.

Reply:
Hmmm... I mean, these types are already ugly; so I'd strive for accuracy.